core/src/main/java/hudson/model/EnvironmentContributingAction.java

```diff
@@ -44,13 +50,42 @@
  * @see BuildWrapper
  */
 public interface EnvironmentContributingAction extends Action {
+    /**
+     * Called by {@link Run} or {@link AbstractBuild} to allow plugins to contribute environment variables.
+     *
+     * @param run
+     *      The calling build. Never null.
+     * @param node
+     *      The node execute on. Can be {@code null} when the Run is not binded to the node,
+     *      e.g. in Pipeline outside the {@code node() step}
+     * @param env
+     *      Environment variables should be added to this map.
+     * @since TODO
+     */
+    default void buildEnvVars(@Nonnull Run<?, ?> run, @Nonnull EnvVars env, @CheckForNull Node node) {
+        if (run instanceof AbstractBuild
+                && Util.isOverridden(EnvironmentContributingAction.class,
+                                     getClass(), "buildEnvVars", AbstractBuild.class, EnvVars.class)) {
+            buildEnvVars((AbstractBuild) run, env);
+        }
+    }
+
     /**
      * Called by {@link AbstractBuild} to allow plugins to contribute environment variables.
      *
+     * @deprecated Use {@link #buildEnvVars(Run, EnvVars, Node)} instead
+     *
      * @param build
      *      The calling build. Never null.
      * @param env
      *      Environment variables should be added to this map.
      */
-    void buildEnvVars(AbstractBuild<?, ?> build, EnvVars env);
+    @Deprecated
+    @Restricted(ProtectedExternally.class)
```

Comment:
Causing me problems in https://github.com/jenkinsci/maven-plugin/pull/109. Why was this necessary?

Reply:
It was supposed to make upgrading plugin maintainers to migrate to new APIs. But I agree that such approach creates issues in Jenkinsfiles testing against newer cores. PCT is not my concern, it should probably ignore `@Restricted` annotations and just report them in the output without failing the entrire build